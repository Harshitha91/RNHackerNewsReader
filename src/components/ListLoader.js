import React from 'react';
import { Dimensions } from 'react-native';
import ContentLoader, { Rect } from 'react-content-loader/native';
const { width, height } = Dimensions.get('window');

const ListLoader = () => (
    <ContentLoader
        speed={2}
        width={'100%'}
        height={70}
        viewBox={`0 0 ${width - width / 30} 70`}
        backgroundColor="#f3f3f3"
        foregroundColor="#ecebeb">
        <Rect
            x="10"
            y="15"
            rx="3"
            ry="3"
            width={(width - width / 10) / 2}
            height="6"
        />
        <Rect x="10" y="30" rx="3" ry="3" width={width / 4} height="6" />
        <Rect
            x="10"
            y="45"
            rx="3"
            ry="3"
            width={width - width / 10}
            height="6"
        />
    </ContentLoader>
);

export default ListLoader;
