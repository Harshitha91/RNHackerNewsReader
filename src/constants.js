export const API = 'https://hacker-news.firebaseio.com/v0/';

export const HOME_SCREEN = 'HOME_SCREEN';
export const COMMENT_SCREEN = 'COMMENT_SCREEN';
