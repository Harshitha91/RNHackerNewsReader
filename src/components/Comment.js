import React, { useEffect, useState } from 'react';
import { View, Text, StyleSheet } from 'react-native';
import moment from 'moment';
import axios from 'axios';
import { API } from '../constants';
import { ListLoader } from '.';

export default function Comment({ id }) {
    const [comment, setComment] = useState(null);

    useEffect(() => {
        let mounted = true;
        async function fetchData() {
            const response = await axios.get(API + `item/${id}.json`);
            if (mounted) {
                setComment(response?.data);
            }
        }
        fetchData();

        return () => (mounted = false);
    }, []);

    return (
        <View style={styles.container}>
            {!comment && <ListLoader />}
            {comment && (
                <View style={{ marginTop: 5, height: 150 }}>
                    <View style={styles.bodyContent}>
                        <View style={styles.user}>
                            <Text testID={'user'} style={styles.titleStyle}>
                                {comment.by ?? ''}
                            </Text>
                        </View>
                        <View style={styles.time}>
                            <Text style={styles.text}>
                                {comment.time
                                    ? moment(comment.time * 1000).fromNow()
                                    : ''}
                            </Text>
                        </View>
                    </View>
                    <View style={styles.body}>
                        <Text
                            testID={'text'}
                            style={styles.bodyText}
                            numberOfLines={10}
                            ellipsizeMode="tail">
                            {comment.text ?? ''}
                        </Text>
                    </View>
                </View>
            )}
        </View>
    );
}
const styles = StyleSheet.create({
    container: {
        alignSelf: 'stretch',
        backgroundColor: '#FFF',
    },
    bodyContent: {
        padding: 5,
        paddingLeft: 15,
        flexDirection: 'row',
        justifyContent: 'flex-start',
        alignItems: 'center',
    },

    body: {
        height: 120,
        paddingLeft: 15,
    },
    bodyText: {
        fontSize: 14,
        lineHeight: 20,
        color: '#424242',
        marginRight: 20,
        paddingRight: 5,
        flex: 1,
    },
    user: {
        flexDirection: 'row',
        width: 111,
        height: 20,
        justifyContent: 'flex-start',
        alignItems: 'center',
    },
    titleStyle: {
        margin: 0,
        padding: 0,
        fontSize: 14,
        color: 'rgba(255,102,13,1)',
        marginLeft: 5,
    },
    time: {
        width: 111,
        height: 20,
        flexDirection: 'row',
        alignItems: 'center',
        justifyContent: 'flex-start',
    },
    text: {
        margin: 0,
        marginLeft: 5,
        padding: 0,
        fontSize: 14,
        color: 'rgba(177,177,177,1)',
    },
});
