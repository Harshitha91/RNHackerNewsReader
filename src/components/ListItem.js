import React, { useEffect, useState } from 'react';
import { View, Text, StyleSheet, TouchableOpacity } from 'react-native';
import axios from 'axios';
import { API, COMMENT_SCREEN } from '../constants';
import { ListLoader } from '.';

export default function ListItem({ id, navigation }) {
    const [newsData, setNewsData] = useState(null);

    useEffect(() => {
        let mounted = true;
        async function fetchData() {
            const response = await axios.get(API + `item/${id}.json`);
            if (mounted) {
                setNewsData(response?.data);
            }
        }
        fetchData();

        return () => (mounted = false);
    }, []);

    return (
        <View style={styles.container}>
            {!newsData && <ListLoader />}
            {newsData && (
                <>
                    <View style={styles.itemContainer}>
                        <Text
                            style={styles.itemName}
                            numberOfLines={2}
                            ellipsizeMode="tail">
                            {newsData?.title}
                        </Text>
                        <Text style={styles.itemAuthor}>{newsData?.by}</Text>
                        <Text
                            style={styles.itemUrl}
                            numberOfLines={2}
                            ellipsizeMode="tail">
                            {newsData?.url}
                        </Text>
                    </View>
                    <TouchableOpacity
                        testID={'button'}
                        style={styles.button}
                        onPress={() => {
                            navigation.push(COMMENT_SCREEN, {
                                commentsList: newsData?.kids ?? [],
                                title: newsData?.title ?? '',
                                url: newsData?.url ?? '',
                            });
                        }}>
                        <View style={styles.commentsContainer}>
                            <Text
                                testID={'comments-number'}
                                style={styles.commentNumbers}>
                                {newsData.kids?.length ?? ''}
                            </Text>
                        </View>
                        <View style={styles.upvotesContainer}>
                            <Text
                                testID={'score-number'}
                                style={styles.upvotes}>
                                {newsData.score ?? ''}
                            </Text>
                        </View>
                    </TouchableOpacity>
                </>
            )}
        </View>
    );
}
const styles = StyleSheet.create({
    container: {
        backgroundColor: '#ffffff',
        height: 90,
        alignSelf: 'stretch',
    },
    itemContainer: {
        top: 0,
        left: 0,
        right: 51,
        height: 90,
        flexDirection: 'column',
        justifyContent: 'space-between',
        backgroundColor: 'rgba(249,249,249,1)',
        opacity: 1,
        paddingRight: 55,
        alignSelf: 'stretch',
    },
    itemName: {
        backgroundColor: 'transparent',
        marginTop: 10,
        marginLeft: 10,
        fontSize: 14,
        color: 'rgba(31,31,31,1)',
    },
    itemAuthor: {
        backgroundColor: 'transparent',
        marginBottom: 0,
        marginLeft: 10,
        fontSize: 12,
        color: 'rgba(148,148,148,1)',
    },
    itemUrl: {
        backgroundColor: 'transparent',
        marginBottom: 10,
        marginLeft: 10,
        fontSize: 12,
        color: 'rgba(148,148,148,1)',
    },
    button: {
        top: 0,
        width: 51,
        height: 90,
        position: 'absolute',
        backgroundColor: 'rgba(245,245,245,1)',
        opacity: 1,
        right: 0,
    },
    commentsContainer: {
        top: 112,
        left: 10,
        width: 32,
        height: 20,
        position: 'absolute',
    },
    commentNumbers: {
        top: -94,
        left: 4,
        position: 'absolute',
        backgroundColor: 'transparent',
        fontSize: 10,
        color: 'rgba(255,102,0,1)',
        width: 24,
        height: 12,
        textAlign: 'center',
    },
    upvotesContainer: {
        top: 54,
        left: 0,
        right: 0,
        height: 20,
        position: 'absolute',
        backgroundColor: 'rgb(230,230,230)',
    },
    upvotes: {
        top: 4,
        left: 11,
        position: 'absolute',
        backgroundColor: 'transparent',
        fontSize: 10,
        textAlign: 'center',
        color: 'rgba(156,156,156,1)',
        width: 30,
        height: 12,
    },
});
