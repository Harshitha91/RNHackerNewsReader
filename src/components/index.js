import ListItem from './ListItem';
import Comment from './Comment';
import ListLoader from './ListLoader';

export { ListItem, Comment, ListLoader };
