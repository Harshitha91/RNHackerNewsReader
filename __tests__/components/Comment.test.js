import * as React from 'react';
import { Comment } from '../../src/components';
import { render, cleanup } from '@testing-library/react-native';
import axios from 'axios';
import renderer from 'react-test-renderer';

jest.mock('axios');

describe('Test Comment Component.', () => {
    beforeAll(() => {
        const fakeResponse = {
            data: {
                by: '3np',
                id: 31501124,
                parent: 31500684,
                text: '(2019).<p>Previous discussion: <a href="https:&#x2F;&#x2F;news.ycombinator.com&#x2F;item?id=21759683" rel="nofollow">https:&#x2F;&#x2F;news.ycombinator.com&#x2F;item?id=21759683</a>',
                time: 1653458809,
                type: 'comment',
            },
        };
        axios.get.mockResolvedValue(fakeResponse);
    });

    afterAll(cleanup);

    it('Should visible fetched users name.', async () => {
        const { findByTestId } = render(<Comment />);
        const user = await findByTestId('user');
        expect(user.children[0]).toBe('3np');
    });

    it('Should update score with fetched data.', async () => {
        const { findByTestId } = render(<Comment />);
        const text = await findByTestId('text');
        expect(text).toBeTruthy();
    });
});
