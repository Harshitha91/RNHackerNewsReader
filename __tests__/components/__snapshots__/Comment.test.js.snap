// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<View
  style={
    Object {
      "alignSelf": "stretch",
      "backgroundColor": "#FFF",
    }
  }
>
  <RNSVGSvgView
    align="xMidYMid"
    bbHeight={70}
    bbWidth="100%"
    focusable={false}
    height={70}
    interval={0.25}
    meetOrSlice={0}
    minX={0}
    minY={0}
    style={
      Array [
        Object {
          "backgroundColor": "transparent",
          "borderWidth": 0,
        },
        Object {},
        Object {
          "flex": 0,
          "height": 70,
          "width": "100%",
        },
      ]
    }
    vbHeight={70}
    vbWidth={725}
    width="100%"
  >
    <RNSVGGroup>
      <RNSVGRect
        clipPath="14bezy-animated-diff"
        fill={
          Array [
            1,
            "14bezy-diff",
          ]
        }
        height="100%"
        propList={
          Array [
            "fill",
          ]
        }
        width="100%"
        x="0"
        y="0"
      />
      <RNSVGDefs>
        <RNSVGClipPath
          name="14bezy-animated-diff"
        >
          <RNSVGRect
            height="6"
            rx="3"
            ry="3"
            width={337.5}
            x="10"
            y="15"
          />
          <RNSVGRect
            height="6"
            rx="3"
            ry="3"
            width={187.5}
            x="10"
            y="30"
          />
          <RNSVGRect
            height="6"
            rx="3"
            ry="3"
            width={675}
            x="10"
            y="45"
          />
        </RNSVGClipPath>
        <RNSVGLinearGradient
          gradient={
            Array [
              0,
              -789517,
              0.5,
              -1250325,
              1,
              -789517,
            ]
          }
          gradientTransform={null}
          gradientUnits={0}
          name="14bezy-diff"
          x1="-100%"
          x2="0%"
          y1={0}
          y2={0}
        />
      </RNSVGDefs>
    </RNSVGGroup>
  </RNSVGSvgView>
</View>
`;
